### ISSUE TITLE
Refresh token error

### ISSUE THREAD

**Issue by dariusdamalakas** [ringcentral/ringcentral-csharp#44]
We need to be able to catch refresh token errors, as that indicates the integration is down until client reauthenticates and we get a new refresh token.

We get a generic `ApiException`:
````
RingCentral.Http.ApiException: refresh token is wrong!
   at RingCentral.Http.ApiResponse..ctor(HttpResponseMessage response, HttpRequestMessage request)
   at RingCentral.Platform.AuthCall(Request request)
   at RingCentral.Platform.Refresh()
````

Preferably, this could be a `RefreshTokenInvalid` exception or similar.

**Comment by dariusdamalakas** [comment]
P.s. am looking at the code what's the best approach here.

**Comment by tylerlong** [comment]
Currently there is only one general exception class named `ApiException`. And the exception reason is included in the exception message. I think most of the time it's enough. That being said, it is nice to have more specific exceptions.

**Comment by dariusdamalakas** [comment]
@tylerlong , yes, that's what I though - i have the error in the message, it might be enough to just look into that.

**Comment by LingForCC** [comment]
The error message associated with Exception is prepared for human. It is pain for the code to deal with exception which only has error message. I think specific exceptions should be provided in future release.

**Comment by dariusdamalakas** [comment]
@LingForCC I guess we need to look case by case:
* `throw new Exception("Refresh Token has Expired");` from [Refresh()](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L75) method
* `throw new Exception("Access has Expired");` from  [Send](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L145)
* ` if (!string.IsNullOrEmpty((string)(data["error_description"])))` From [ApiResponse.Error property](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Http/ApiResponse.cs#L112)

Actually, slightly related is the error `429` which I am pretty sure we will hitting. We'd like to be able to properly catch this, look up into the header and suspend our operations. So thats another thing to consider.

**Comment by dariusdamalakas** [comment]
Actually, the biggest issue I see with current exception handling is that generic `Exception` objects are thrown, instead of specific. This essentially prevents us from capturing genuine issues that we might see, and instead we have to provide a _catch-all_ clause, to catch refresh token errors.  This must definitely be addressed. Options:

1. Throw `ApiException` would be better. Better than throwin `Exception`
1. Throw `AuthException`
1. Create hierarchy `AuthException` and 2 subclasses: `RefreshTokenInvalid` and `AccessTokenInvalid`.

I'm leaning towards the third option.

**Comment by tylerlong** [comment]
quick question: `AccessTokenInvalid`, sdk will try to Refresh if refreshToken is valid. https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L222

Is it necessary to throw exception in such case?

**Comment by dariusdamalakas** [comment]
@tylerlong, good point, actually no, it would not make sense to throw `AccessTokenInvalid` if refresh token is still valid. So maybe perhaps just throwing `AuthException` would be enough. 

By the way, we still need to incorporate into the "grand design" how we can deal with 429 exceptions.

**Comment by dariusdamalakas** [comment]
Several options regarding to exception handling 
1. Don't throw exception if non 20x resposne returned - move this responsibility to client. Breaking change.
1. Throw exceptions for each type of http code - overkill to implement
1. Throw `ApiException` but pass in response and request message.

IMO, the third option is least change, and aligns best with what you currently do. Plus with C#6 exception filter feature, clients will be able to handle different statuses nicely (or you can do that with an `if` statement now)

```` 
try {
  // Do stuff
} 
catch (ApiException e) when (e.Response.Status == HttpStatus.TooManyRequests) { // handle this }
catch (ApiException e) when (e.Response.Status == HttpStatus.Forbidden) { // handle this }
````

Actually, if a `refresh_token` is expired, which `HTTP status` is that?

**Comment by tylerlong** [comment]
Good ideas. I agree that #3 is the easiest to implement. Maybe I will do it on Monday. 

> Actually, if a refresh_token is expired, which HTTP status is that?

I think it is the same as an invalid refresh_token, so you can have a try.

Another thing is: actually we don't talk to the server when checking refresh_token. We just compare the expiration time when current time. So there is no HTTP. I am not sure if this is the ideal way.  https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Auth.cs#L222

Maybe we just maintain two exceptions: `AuthException` (which doesn't talk to server) and `ApiException` (which talks to server and response is not `20x`, request and response message will be passed in).

**Comment by dariusdamalakas** [comment]
@tylerlong Btw, just to be clear, we don't need the fix ASAP for this, we have forked your repo, and published our own nuget. But if you do make changes, we can just pull them from whatever branch you are making these and test them out. So no rush from our side, we've got things moving on our end.

> We just compare the expiration time with current time. So there is no HTTP. I am not sure if this is the ideal way. 

a) RingCentral also implements [token revocation](https://developer.ringcentral.com/api-docs/latest/index.html#!#RefRevokeToken.html). That is, an `access_token` and `refresh_token` can be revoked by ring central authorization server. In that case, checking only tokens TTL values will not tell whether token is still valid or not. 
b) Using `refresh_token` to obtain new token pairs invalides previously issued tokens ([see docs](https://developer.ringcentral.com/api-docs/latest/index.html#!#RefRefreshTokenFlow)). This actually does have an interesting implication about how RC implements resource servers - i.e. each operation needs to call authorization server to validate bearer token. This is a huge performance implication, but perhaps there are some architectural decisions that required this.

So basically checking TTL to determine if any of the token is still valid is not enough.

**Comment by tylerlong** [comment]
I updated `ApiException` in this commit so it contains `ApiResponse`.  Request object is not contained by exception object because it is already contained by response object: `apiException.response.request`.

**Comment by tylerlong** [comment]
`ApiException` is provided with version `1.0.1`. You can get detailed request and response information by inspect the exception instance. So it should be sufficient for programmers to do troubleshooting.

For authorization exceptions. I didn't create any special exceptions classes. Because as long as the authorization expires new requests will be rejected by server and thus you can get an exception to investigate.

Close it for now. Feel free to reopen it or create a new issue.
